### ISSUE TITLE
Uncaught Range Error: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by StlthyLee** [atom/atom#13181]
[Enter steps to reproduce below:]

1. open any file (in my case JS) 
2. select any text to delete, it doesn't matter if it's three characters or thirty I still get the issue.

**Atom Version**: 1.12.0
**Electron Version**: 1.3.6
**System**: Mac OS X 10.12.1
**Thrown From**: Atom Core

### Stack Trace

Uncaught RangeError: Maximum call stack size exceeded

```
At /Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:558

RangeError: Maximum call stack size exceeded
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:558:7)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
    at KeymapManager.module.exports.KeymapManager.terminatePendingState (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:565:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:423:21)
```

### Commands

```
     -4:26.6.0 atomic-emacs:backward-char (atom-text-editor.editor.is-focused)
     -4:25.8.0 atomic-emacs:forward-char (atom-text-editor.editor.is-focused)
  2x -4:25.4.0 atomic-emacs:backward-char (atom-text-editor.editor.is-focused)
     -4:20.6.0 core:save (atom-text-editor.editor.is-focused.autocomplete-active)
     -3:52 platformio-ide-terminal:copy (span.)
  4x -1:54.1.0 core:backspace (atom-text-editor.editor.is-focused)
     -1:49.1.0 core:save (atom-text-editor.editor.is-focused)
     -1:20 editor:toggle-line-comments (atom-text-editor.editor.is-focused)
     -1:20 core:select-all (atom-text-editor.editor.is-focused)
     -1:16.2.0 editor:toggle-line-comments (atom-text-editor.editor.is-focused)
     -1:15.3.0 core:save (atom-text-editor.editor.is-focused)
     -0:38.2.0 core:backspace (atom-text-editor.editor.is-focused)
     -0:36.1.0 core:save (atom-text-editor.editor.is-focused)
  4x -0:34 core:backspace (atom-text-editor.editor.is-focused)
     -0:28.8.0 core:save (atom-text-editor.editor.is-focused)
     -0:00.0 core:copy (atom-text-editor.editor.is-focused)
```

### Config

```json
{
  "core": {
    "disabledPackages": [
      "linter-jscs",
      "jshint"
    ],
    "telemetryConsent": "no",
    "themes": [
      "one-dark-ui",
      "monokai"
    ]
  },
  "editor": {
    "fontFamily": "Inconsolata",
    "fontSize": 16,
    "preferredLineLength": 120,
    "showIndentGuide": true,
    "softWrapAtPreferredLineLength": true,
    "tabType": "soft"
  }
}
```

### Installed Packages

```coffee
# User
atom-bootstrap3, v1.2.12 (inactive)
atom-snippets, v0.1.1 (active)
atomic-emacs, v0.9.2 (active)
es6-javascript, v1.0.0 (inactive)
gaia-dark-syntax, v1.1.0 (inactive)
html-id-class-snippets, v1.4.1 (active)
jshint, v1.8.6 (inactive)
linter, v1.11.18 (active)
linter-eslint, v8.0.0 (active)
linter-htmlhint, v1.3.1 (active)
linter-jscs, v4.1.0 (inactive)
monokai, v0.18.0 (active)
platformio-ide-terminal, v2.2.1 (active)
railscast-theme, v0.3.0 (inactive)
railscasts, v0.2.2 (inactive)
react, v0.16.1 (active)
react-es6-snippets, v0.3.0 (active)
atom-dark-syntax, v0.27.0 (inactive)
atom-dark-ui, v0.52.0 (inactive)
atom-light-syntax, v0.28.0 (inactive)
atom-light-ui, v0.45.0 (inactive)
base16-tomorrow-dark-theme, v1.3.0 (inactive)
base16-tomorrow-light-theme, v1.3.0 (inactive)
one-dark-ui, v1.6.2 (active)
one-light-ui, v1.6.2 (inactive)
one-dark-syntax, v1.5.0 (inactive)
one-light-syntax, v1.5.0 (inactive)
solarized-dark-syntax, v1.0.5 (inactive)
solarized-light-syntax, v1.0.5 (inactive)
about, v1.7.0 (active)
archive-view, v0.62.0 (active)
autocomplete-atom-api, v0.10.0 (active)
autocomplete-css, v0.13.1 (active)
autocomplete-html, v0.7.2 (active)
autocomplete-plus, v2.31.4 (active)
autocomplete-snippets, v1.11.0 (active)
autoflow, v0.27.0 (inactive)
autosave, v0.23.1 (active)
background-tips, v0.26.1 (active)
bookmarks, v0.42.0 (active)
bracket-matcher, v0.82.2 (active)
command-palette, v0.39.0 (inactive)
deprecation-cop, v0.54.1 (active)
dev-live-reload, v0.47.0 (active)
encoding-selector, v0.22.0 (active)
exception-reporting, v0.40.0 (active)
find-and-replace, v0.202.0 (inactive)
fuzzy-finder, v1.4.0 (active)
git-diff, v1.1.0 (active)
go-to-line, v0.31.0 (inactive)
grammar-selector, v0.48.2 (active)
image-view, v0.60.0 (active)
incompatible-packages, v0.26.1 (active)
keybinding-resolver, v0.35.0 (active)
line-ending-selector, v0.5.0 (active)
link, v0.31.2 (inactive)
markdown-preview, v0.158.8 (active)
metrics, v1.1.1 (active)
notifications, v0.65.1 (active)
open-on-github, v1.2.1 (inactive)
package-generator, v1.0.1 (inactive)
settings-view, v0.243.1 (active)
snippets, v1.0.3 (active)
spell-check, v0.68.4 (active)
status-bar, v1.4.1 (active)
styleguide, v0.47.2 (active)
symbols-view, v0.113.1 (inactive)
tabs, v0.103.0 (active)
timecop, v0.33.2 (active)
tree-view, v0.210.0 (active)
update-package-dependencies, v0.10.0 (active)
welcome, v0.35.1 (active)
whitespace, v0.35.0 (active)
wrap-guide, v0.38.2 (active)
language-c, v0.54.0 (active)
language-clojure, v0.22.1 (active)
language-coffee-script, v0.48.0 (active)
language-csharp, v0.12.1 (active)
language-css, v0.40.1 (active)
language-gfm, v0.88.0 (active)
language-git, v0.15.0 (active)
language-go, v0.43.0 (active)
language-html, v0.46.1 (active)
language-hyperlink, v0.16.1 (active)
language-java, v0.24.0 (active)
language-javascript, v0.122.0 (active)
language-json, v0.18.3 (active)
language-less, v0.29.6 (active)
language-make, v0.22.2 (active)
language-mustache, v0.13.0 (active)
language-objective-c, v0.15.1 (active)
language-perl, v0.37.0 (active)
language-php, v0.37.3 (active)
language-property-list, v0.8.0 (active)
language-python, v0.45.1 (active)
language-ruby, v0.70.2 (active)
language-ruby-on-rails, v0.25.1 (active)
language-sass, v0.57.0 (active)
language-shellscript, v0.23.0 (active)
language-source, v0.9.0 (active)
language-sql, v0.25.0 (active)
language-text, v0.7.1 (active)
language-todo, v0.29.1 (active)
language-toml, v0.18.1 (active)
language-xml, v0.34.12 (active)
language-yaml, v0.27.1 (active)

# Dev
No dev packages
```

**Comment by Ben3eeE** [comment]
Can you reproduce this in safe mode `atom --safe`?

**Comment by StlthyLee** [comment]
No I can't seem to reproduce the issue in safe mode

**Comment by 50Wliu** [comment]
Are you by any chance using the new keymap API?  If so please paste the contents of your `init.coffee` file here.

**Comment by StlthyLee** [comment]
@50Wliu No I'm not using the new keymap API.

**Comment by joshmanders** [comment]
Getting the same issue and I haven't done anything but update Atom and rebuild incompatible packages.

**Comment by 50Wliu** [comment]
@joshmanders Can you reproduce this issue in safe mode (`atom --safe`)?

**Comment by joshmanders** [comment]
Unfortunately I can't test, I am in the middle of important work so I had reverted back to `1.11.2`.

**Comment by maxbrunsfeld** [comment]
Thanks for the reports! Can everyone who is experiencing this share their `keymap.cson`?

**Comment by nathansobo** [comment]
If you can figure out what package introduces this issue we will be able to address it. That's obviously annoying but unless we can reproduce this it's really hard to know what to address.

**Comment by StlthyLee** [comment]
Unfortunately I haven't been able to pin this down to a specific package yet, especially as the error is intermittent. I will close this for now but reopen if I do manage to find a way to reproduce the problem.

Thanks.

**Comment by Ben3eeE** [comment]
@StlthyLee Do you have anything in your `keymap.cson`?

**Comment by Ben3eeE** [comment]
@nathansobo @ungb We have 2 other reports in (wrong issue) https://github.com/atom/atom/issues/11560. Seems all reports where we have information use `platformio-ide-terminal`, `linter`, `linter-eslint` and macOS in common.

Can one of these packages cause this? Is anyone that experience this problem willing to test this?

**Comment by nathansobo** [comment]
@ungb This would be a really helpful one to reproduce if possible.

**Comment by grahamb** [comment]
I'm one of the reports from #11560. I had commented everything out of `keymap.cson` but was still getting the issue. I launched atom in safe mode a few hours ago and it's been stable, so I'd agree it's one of those packages. I can try enabling them one at a time to see if I can reproduce.

**Comment by izaera** [comment]
In response to @Ben3eeE request, I'm adding info copied from #6837 to this bug (please review my older comments in that bug to get more context info).

---
I'm attaching a console log to this comment with the relevant information. Is not definitive, but it appears that KeymapManager is remembering past keystrokes that didn't get executed and gets fooled because of that, entering an infinite loop/recursion.

If you see the log from the end, the keypress resolving algorithm keeps processing Meta, Enter, t, and k forever. This is probably due to the fact that I have a terminal plugin installed and, sometimes, I press Meta+K to clear its output.

This works when the terminal has the focus, but sometimes a single click is not enough and the Meta+K goes to atom and is remembered forever, until I go to an editor, press an arrow key, and atom interprets that as a window split (and it fact it does the split).

I believe those two events: the hang and the focus thing are related and somehow interfering with atom's KeymapManager.

Said that, I'm not sure how to continue. Would it be possible that someone from the dev team familiarized with the KeymapManager review this and tell me what more info I can gather to help fixing the problem?

I have tried to understand KeymapManager, but I don't have plenty of time, and it looks like a difficult algorithm to patch it without understanding.

Thx.

[log_error_atom.txt](https://github.com/atom/atom/files/599503/log_error_atom.txt)

---
One more piece of info:

This bug always showed the stack overflow notification when it happened, but now it is hanging Atom (no responsive UI, though menu is still working, devtools too, and the window repaints). Looks like a deadlock, because CPU is at 0%.

Don't know why this change in behavior. It could be due to the traces I have added, but I'm not sure.

**Comment by Ben3eeE** [comment]
@izaera Are you using `platformio-ide-terminal`?

**Comment by izaera** [comment]
Yes.

**Comment by StlthyLee** [comment]
I believe the issue is caused by platformio-ide-terminal as until a couple of days ago I was using it, and once I stopped I was no longer seeing this issue

**Comment by izaera** [comment]
I have reviewed all comments in this bug and wanted to note that I do have `linter` and `linter-eslint` too (in addtion to Mac and `platformio-ide-terminal`, of course). 

But I believe too, that if a package is the culprit, it is most probably the terminal, because I had the same issue again during this morning and the Meta and K keys were involved again.

**Comment by izaera** [comment]
Good news guys :-). I can reproduce it \o/ .

You need to do this:
1) install `platformio-ide-terminal`
2) click on the + button of terminal to create a new terminal window and leave it open (don't minimize it)
3) open an editor (any file is ok)
4) focus the terminal and hit Cmd+K (this will clear the terminal)
5) click with the mouse on the editor (this will focus the editor and put a blinking caret on it)
6) now, switch to other app (with Cmd+Tab)
7) click on atom's window once, inside the terminal area (this will focus atom, but leave the blinking caret on the editor; as it was before)
8) hit Cmd+K and it will hang

I suggest not maximizing atom to ease step 7 because you must play with the mouse and focus, if you switch back to atom with Cmd+Tab it does not happen. 

Historical note: the bug hit me several times because I work with two monitors, and click several times on Atom to switch back to it, instead of using Cmd+Tab. Also, I usually hit Cmd+K after clicking on atom because after testing something in the browser I come back to atom and clean the output of rails in the terminal. Otherwise it would have been difficult to get hit by this bug so often.

**Comment by nathansobo** [comment]
@izaera Awesome work. I'm now reproducing this via your steps and investigating the root cause.

**Comment by iolsen** [comment]
Fix shipped today in 1.12.5.
